File: OCEL.CSharp.Tests/UtilityTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;
using Xunit.Abstractions;

namespace OCEL.CSharp.Tests
{
    public class UtilityTests
    {
        private readonly ITestOutputHelper _output;

        public UtilityTests(ITestOutputHelper output)
        {
            _output = output;
        }

        [Fact]
        public void CanCorrectlyMergeTwoLogs()
        {
            var log1 = new OcelLog(
                new Dictionary<string, OcelValue>
                {
                    { "ordering", new OcelString("timestamp") },
                },
                new Dictionary<string, OcelEvent>
                {
                    { "e1", new OcelEvent("Activity 1", DateTimeOffset.Now, new List<string> { "o1" }, new Dictionary<string, OcelValue>()) },
                    { "e2", new OcelEvent("Activity 2", DateTimeOffset.Now.AddDays(1), new List<string> { "o1" }, new Dictionary<string, OcelValue>()) }
                },
                new Dictionary<string, OcelObject>
                {
                    { "o1", new OcelObject("Object", new Dictionary<string, OcelValue>())}
                }
            );

            var log2 = new OcelLog(
                new Dictionary<string, OcelValue>
                {
                    { "ordering", new OcelString("timestamp") },
                    { "version", new OcelString("1.0") }
                },
                new Dictionary<string, OcelEvent>
                {
                    { "e1", new OcelEvent("Activity 1", DateTimeOffset.Now, new List<string> { "o1" }, new Dictionary<string, OcelValue>()) },
                    { "e3", new OcelEvent("Activity 2", DateTimeOffset.Now.AddDays(1), new List<string> { "o1" }, new Dictionary<string, OcelValue>()) }
                },
                new Dictionary<string, OcelObject>
                {
                    { "o1", new OcelObject("Object", new Dictionary<string, OcelValue>())}
                }
            );

            var merged = log1.MergeWith(log2);

            Assert.True(merged.GlobalAttributes.Count == 2);
            Assert.True(merged.GlobalAttributes.Keys.SequenceEqual(new List<string> { "ordering", "version" }));

            Assert.True(merged.Events.Count == 3);
            Assert.True(merged.Events.Keys.SequenceEqual(new List<string> { "e1", "e2", "e3" }));

            Assert.True(merged.Objects.Count == 1);
            Assert.True(merged.Objects.Keys.SequenceEqual(new List<string> { "o1" }));
        }

        [Fact]
        public void CanCorrectlyRemoveDuplicateObjects()
        {
            var timestamp = DateTimeOffset.Now;
            var log = new OcelLog(
                new Dictionary<string, OcelValue>(),
                new Dictionary<string, OcelEvent>
                {
                    { "e1", new OcelEvent("Activity 1", timestamp, new List<string> { "o1" }, new Dictionary<string, OcelValue>()) },
                    { "e2", new OcelEvent("Activity 2", timestamp, new List<string> { "o1-duplicate" }, new Dictionary<string, OcelValue>()) }
                },
                new Dictionary<string, OcelObject>
                {
                    { "o1", new OcelObject("Object", new Dictionary<string, OcelValue> { { "test", new OcelString("some string") }, { "test2", new OcelInteger(123) } })},
                    { "o1-duplicate", new OcelObject("Object", new Dictionary<string, OcelValue> { { "test", new OcelString("some string") }, { "test2", new OcelInteger(123) } })},
                }
            );

            var correctLog = new OcelLog(
                new Dictionary<string, OcelValue>(),
                new Dictionary<string, OcelEvent>
                {
                    { "e1", new OcelEvent("Activity 1", timestamp, new List<string> { "o1" }, new Dictionary<string, OcelValue>()) },
                    { "e2", new OcelEvent("Activity 2", timestamp, new List<string> { "o1" }, new Dictionary<string, OcelValue>()) }
                },
                new Dictionary<string, OcelObject>
                {
                    { "o1", new OcelObject("Object", new Dictionary<string, OcelValue> { { "test", new OcelString("some string") }, { "test2", new OcelInteger(123) } })}
                }
            );

            Assert.True(log.MergeDuplicateObjects() == correctLog);
        }

        [Fact]
        public void CanCorrectlyConvertObjectsToAttributes()
        {
            var timestamp = DateTimeOffset.Now;
            var log = new OcelLog(
                new Dictionary<string, OcelValue>(),
                new Dictionary<string, OcelEvent>
                {
                    { "e1", new OcelEvent("Activity 1", timestamp, new List<string> { "o1", "o2" }, new Dictionary<string, OcelValue>()) },
                    { "e2", new OcelEvent("Activity 2", timestamp, new List<string> { "o2" }, new Dictionary<string, OcelValue>()) }
                },
                new Dictionary<string, OcelObject>
                {
                    { "o1", new OcelObject("Type 1", new Dictionary<string, OcelValue> { { "prop1", new OcelString("some string") }, { "prop2", new OcelInteger(123) } })},
                    { "o2", new OcelObject("Type 2", new Dictionary<string, OcelValue> { { "prop1", new OcelBoolean(false) } })},
                }
            );

            var correctLog = new OcelLog(
                new Dictionary<string, OcelValue>(),
                new Dictionary<string, OcelEvent>
                {
                    { "e1", new OcelEvent("Activity 1", timestamp, new List<string>(), 
                        new Dictionary<string, OcelValue>
                        {
                            { "Type 1", new OcelMap(new Dictionary<string, OcelValue> {
                                { "prop1", new OcelString("some string") },
                                { "prop2", new OcelInteger(123) }
                            })},
                            { "Type 2", new OcelBoolean(false) }
                        })
                    },
                    { "e2", new OcelEvent("Activity 2", timestamp, new List<string>(), new Dictionary<string, OcelValue>
                    {
                        { "Type 2", new OcelBoolean(false) }
                    })}
                },
                new Dictionary<string, OcelObject>()
            );

            Assert.True(log.ConvertObjectsToAttributes(new List<string> { "Type 1", "Type 2" }) == correctLog);
        }

        [Fact]
        public void CanCorrectlyConvertAttributesToObjects()
        {
            var timestamp = DateTimeOffset.Now;
            var log = new OcelLog(
                new Dictionary<string, OcelValue>(),
                new Dictionary<string, OcelEvent>
                {
                    { "e1", new OcelEvent("Activity 1", timestamp, new List<string>(),
                        new Dictionary<string, OcelValue>
                        {
                            { "Type 1", new OcelMap(new Dictionary<string, OcelValue> {
                                { "prop1", new OcelString("some string") },
                                { "prop2", new OcelInteger(123) }
                            })},
                            { "Type 2", new OcelBoolean(false) }
                        })
                    },
                    { "e2", new OcelEvent("Activity 2", timestamp, new List<string>(), new Dictionary<string, OcelValue>
                    {
                        { "Type 2", new OcelBoolean(false) }
                    })}
                },
                new Dictionary<string, OcelObject>()
            );

            var result = log.ConvertAttributesToObjects(new List<string> { "Type 1", "Type 2" });
            Assert.All(result.Events, e =>
            {
                switch (e.Key)
                {
                    case "e1":
                        Assert.True(e.Value.OMap.Count() == 2 && e.Value.VMap.Count == 0);
                        return;
                    case "e2":
                        Assert.True(e.Value.OMap.Count() == 1 && e.Value.VMap.Count == 0);
                        return;
                    default:
                        throw new Exception("Unknown event");
                }
            });
            Assert.True(result.Objects.Count == 2);
        }
    }
}

File: OCEL.Tests/UtilityTests.fs
namespace OCEL.Tests

open OCEL
open OCEL.Types

open System
open Xunit
open Xunit.Abstractions

type UtilityTests(output: ITestOutputHelper) =

    [<Fact>]
    let ``Can correctly merge two logs`` () =
        let log1 = {
            GlobalAttributes = ["ordering", OcelString "timestamp"] |> Map.ofList
            Events = [
                "e1", {
                    Activity = "Activity 1"
                    Timestamp = DateTimeOffset.Now
                    OMap = ["o1"]
                    VMap = Map.empty 
                }
                "e2", {
                    Activity = "Activity 2"
                    Timestamp = DateTimeOffset.Now.AddDays 1
                    OMap = ["o1"]
                    VMap = Map.empty
                }
            ] |> Map.ofList
            Objects = [
                "o1", {
                    Type = "Object"
                    OvMap = Map.empty
                }
            ] |> Map.ofList
        }

        let log2 = {
            GlobalAttributes = ["ordering", OcelString "timestamp"; "version", OcelString "1.0"] |> Map.ofList
            Events = [
                "e1", {
                    Activity = "Activity 1"
                    Timestamp = DateTimeOffset.Now
                    OMap = ["o1"]
                    VMap = Map.empty 
                }
                "e3", {
                    Activity = "Activity 2"
                    Timestamp = DateTimeOffset.Now.AddDays 1
                    OMap = ["o1"]
                    VMap = Map.empty
                }
            ] |> Map.ofList
            Objects = [
                "o1", {
                    Type = "Object"
                    OvMap = Map.empty
                }
            ] |> Map.ofList
        }

        let merged = log1.MergeWith log2

        merged.GlobalAttributes.Count = 2 |> Assert.True
        merged.GlobalAttributes.Keys |> Seq.toList = ["ordering"; "version"] |> Assert.True

        merged.Events.Count = 3 |> Assert.True
        merged.Events.Keys |> Seq.toList = ["e1"; "e2"; "e3"] |> Assert.True

        merged.Objects.Count = 1 |> Assert.True
        merged.Objects.Keys |> Seq.toList = ["o1"] |> Assert.True

    [<Fact>]
    let ``Can correctly remove duplicate objects`` () =
        let timestamp = DateTimeOffset.Now

        let log = {
            GlobalAttributes = Map.empty
            Events = [
                "e1", {
                    Activity = "Activity 1"
                    Timestamp = timestamp
                    OMap = ["o1"]
                    VMap = Map.empty 
                }
                "e2", {
                    Activity = "Activity 2"
                    Timestamp = timestamp
                    OMap = ["o1-duplicate"]
                    VMap = Map.empty
                }
            ] |> Map.ofList
            Objects = [
                "o1", {
                    Type = "Object"
                    OvMap = ["test", OcelString "some string"; "test2", OcelInteger 123] |> Map.ofList
                }
                "o1-duplicate", {
                    Type = "Object"
                    OvMap = ["test", OcelString "some string"; "test2", OcelInteger 123] |> Map.ofList
                }
            ] |> Map.ofList
        }

        let correctLog = {
            GlobalAttributes = Map.empty
            Events = [
                "e1", {
                    Activity = "Activity 1"
                    Timestamp = timestamp
                    OMap = ["o1"]
                    VMap = Map.empty 
                }
                "e2", {
                    Activity = "Activity 2"
                    Timestamp = timestamp
                    OMap = ["o1"]
                    VMap = Map.empty
                }
            ] |> Map.ofList
            Objects = [
                "o1", {
                    Type = "Object"
                    OvMap = ["test", OcelString "some string"; "test2", OcelInteger 123] |> Map.ofList
                }
            ] |> Map.ofList
        }

        log.MergeDuplicateObjects() = correctLog |> Assert.True

    [<Fact>]
    let ``Can correctly convert objects to attributes`` () =
        let timestamp = DateTimeOffset.Now
        let log = {
            GlobalAttributes = Map.empty
            Events = [
                "e1", {
                    Activity = "Activity 1"
                    Timestamp = timestamp
                    OMap = ["o1"; "o2"]
                    VMap = Map.empty 
                }
                "e2", {
                    Activity = "Activity 2"
                    Timestamp = timestamp
                    OMap = ["o2"]
                    VMap = Map.empty
                }
            ] |> Map.ofList
            Objects = [
                "o1", {
                    Type = "Type 1"
                    OvMap = ["prop1", OcelString "some string"; "prop2", OcelInteger 123] |> Map.ofList
                }
                "o2", {
                    Type = "Type 2"
                    OvMap = ["prop1", OcelBoolean false] |> Map.ofList
                }
            ] |> Map.ofList
        }

        let correctLog = {
            GlobalAttributes = Map.empty
            Events = [
                "e1", {
                    Activity = "Activity 1"
                    Timestamp = timestamp
                    OMap = []
                    VMap = [
                        "Type 1", OcelMap(["prop1", OcelString "some string"; "prop2", OcelInteger 123] |> Map.ofList)
                        "Type 2", OcelBoolean false
                    ] |> Map.ofList
                }
                "e2", {
                    Activity = "Activity 2"
                    Timestamp = timestamp
                    OMap = []
                    VMap = ["Type 2", OcelBoolean false] |> Map.ofList
                }
            ] |> Map.ofList
            Objects = Map.empty
        }

        log.ConvertObjectsToAttributes ["Type 1"; "Type 2"] = correctLog |> Assert.True

    [<Fact>]
    let ``Can correctly convert attributes to objects`` () =
        let timestamp = DateTimeOffset.Now
        let log = {
            GlobalAttributes = Map.empty
            Events = [
                "e1", {
                    Activity = "Activity 1"
                    Timestamp = timestamp
                    OMap = []
                    VMap = [
                        "Type 1", OcelMap(["prop1", OcelString "some string"; "prop2", OcelInteger 123] |> Map.ofList)
                        "Type 2", OcelBoolean false
                    ] |> Map.ofList
                }
                "e2", {
                    Activity = "Activity 2"
                    Timestamp = timestamp
                    OMap = []
                    VMap = ["Type 2", OcelBoolean false] |> Map.ofList
                }
            ] |> Map.ofList
            Objects = Map.empty
        }
        
        // Resulting log will have randomly assigned object ID's, so direct equality won't work like in previous tests
        let result = log.ConvertAttributesToObjects ["Type 1"; "Type 2"]
        result.Events |> Map.iter (fun k e -> 
            match k with
            | "e1" -> (e.OMap.Length = 2 && e.VMap.IsEmpty) |> Assert.True
            | "e2" -> (e.OMap.Length = 1 && e.VMap.IsEmpty) |> Assert.True
            | _ -> failwith "Unknown event")
        result.Objects.Count = 2 |> Assert.True
